### PR TITLE
Refactor WindowOperatorQueryKit to use WindowStage class for representing different window stages

### PR DESCRIPTION
**Pull request by Akshat-Jain**
### Description

This PR refactors `WindowOperatorQueryKit` to use a new class `WindowStage` to represent the different window stages. Previously we were dealing with `List<OperatorFactory>` or `List<List<OperatorFactory>>`, and doing a bunch of `instanceof` checks, so this refactor makes it much more readable and cleaner.

This PR does not have any functional change.

<hr>

This PR has:

- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:166]: what's this supposed to fix?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:187]: seems like a possibly problematic way to obtain the rowsignature
it seems to me that `Processor` falls short by returning just `outputColumnNames`

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:166]: Nothing. It's the same code, just with one less nested loop.

Previously:
```java
      for (int i = 0; i < operatorList.size(); i++) {
         for (OperatorFactory operatorFactory : operatorList.get(i)) {
           if (operatorFactory instanceof WindowOperatorFactory) {
```

Now:
```java
      for (int i = 0; i < stages.size(); i++) {
         for (WindowOperatorFactory operatorFactory : stages.get(i).getWindowOperatorFactories()) {
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:238]: To be able to move this logic into the WindowStage class, the WindowStage would have to store some previous state IIUC.

Without previous state, we cannot make the decision of whether a NaiveSortOperator or a NaivePartitioningOperator should go into the same stage or create a new stage. (I'm also trying to avoid making any functional changes in this PR, and keeping it purely a refactoring change)

For example, for queries like
```sql
select countryName, cityName,
count(*) over(partition by countryName order by cityName) as c1,
row_number() over(partition by countryName, cityName order by cityName) as c2
from wikipedia
where cityName in ('New York', 'Aarhus')
group by countryName, cityName;
```

MSQ gets the list of operator factories as:
```
NaiveSortOperatorFactory(d1)
NaivePartitioningOperatorFactory(d1)
WindowOperatorFactory
NaivePartitioningOperatorFactory(d1,d0)
WindowOperatorFactory
```

So we need the previous state to determine whether to add `NaivePartitioningOperatorFactory(d1,d0)` to the current stage or if we need to create a new stage.

Also, I think we'd get a better picture of how this would/should look as part of the GlueingPartitioningOperator PR, so I aimed to keep it simple for this PR. Hope that works!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:379]: > the implementation of this method implicitly makes this class not really relevant

We do need a way to get the final list of operatorFactories for a window stage though? And since we aren't storing a collective `List<OperatorFactory> operatorFactories` in the WindowStage class, we need a place to put them in order.

Please let me know what I'm missing here, thanks!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: Done

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: Done

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:235]: Does the following logic look good for a method `WindowStage#canAccept()`?

```java
public boolean canAccept(OperatorFactory operatorFactory)
    {
      if (operatorFactory instanceof NaiveSortOperatorFactory) {
        return false; // Start a new stage
      }

      if (operatorFactory instanceof WindowOperatorFactory) {
        return true; // Add to current stage
      }

      // If it's a PartitioningOperatorFactory, we can add it to current stage if NaiveSortOperatorFactory is present.
      return sortOperatorFactory != null;
    }
```

(PS: I'm trying to avoid making any functional changes in this PR, and keeping it purely a refactoring change)

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:163]: this seems dodgy...regardless when it got here
I mean it looks up the type of an `outputColumn` from the *inputSignature* ...
it *may* be the same but I don't think there should be gurantees for that...

anyhow...this part should live as `Stage#getRowSignature` or something....

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:235]: it could be a start - but its buggy as it will never accept a `NaiveSortOperatorFactory` ?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:340]: for now we may have to keep this; but instead of the `sortOperatorFactory` and `partitioningOperatorFactory` - this is the place where those could be replaced with the `GlueingPartitionOperator`

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:212]: please file a PR to remove the `query` from all of these classes; they shouldn't need it - its confusing to pass that
if something is need from the `context()` flatten that into a configuration class or something and pass that

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:211]: I feel like the full stage should be build by the `WindowStage`

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (kgyrtkirk) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:400]: list the fields of the class and not shortcut with `getOperatorFactories` ; this class shouldn't be just a fancy list

**Comment by Akshat-Jain** [comment]
@kgyrtkirk Thank you for the review!

I was wondering if it makes sense to get https://github.com/apache/druid/pull/17038 merged first, and then resolve conflicts on this PR, instead of the other way round? Plus this PR is much smaller than the other one, so it would allow us to iterate faster on this after getting the other PR merged.

If we try to merge this PR first, a lot of this PR's logic would have to be re-worked when resolving conflicts on the GlueingPartitioningOperator PR. That PR is already huge, it would be unnecessary to re-work this PR's logic there.

Thoughts?

**Comment by kgyrtkirk** [comment]
we will merge the PR which is ready first

I've suggested to open this from that pr [here](https://github.com/apache/druid/pull/17038#discussion_r1775555695) - so I think the natural order is to have this one finished first

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:235]: Currently, we start a new stage immediately after every "consecutive chain of WindowOperatorFactory".

This effectively means that we always start a new stage in 2 occasions:
1. On encountering NaiveSortOperatorFactory
2. On encountering NaivePartitioningOperatorFactory, but not having encountered a NaiveSortOperatorFactory

The code snippet I posted above represents this logic. (and it works, all tests pass)

This logic would have to be completely re-worked with the GlueingPartitioningOperator PR changes (I don't know how that will look like), which is why I was hoping to get the other PR merged first.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:163]: > anyhow...this part should live as Stage#getRowSignature or something....

I tried adding WindowStage#getRowSignature. But it's not clean, we have to pass a bunch of params, and makes the code difficult to understand. So I feel we should stick with the existing code as it's cleaner and easier to follow, and we should keep WindowStage limited to the operator factory handling.

image

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:212]: Will take this up later as it is unrelated and shouldn't block this PR, hope that's okay.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:184]: I added this in https://github.com/apache/druid/pull/16729/files#diff-1cdba5d733d1ae401b38b93e100543b648fe602e2e434c3e4a4ed23e877a7d66, we had an offline discussion on this at the time of that PR where we agreed and finalized on this approach 😅

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: Have made the change.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: > this is the place where those could be replaced with the GlueingPartitionOperator

Yeah

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: I've pushed a change to create a `WindowStages` class which encapsulates `List<WindowStage>` along with some other properties. `WindowStage` wasn't self-sufficient for calculating everything for the window stage, since we rely on `window stage "x + 1"` for the shuffle spec of `window stage "x"`.

Does this approach work?

> to clean that up you need to enhance the stagebuilder

Replacing `queryDefBuilder.add(work + shuffle)` with `queryDefBuilder.add(work) + queryDefBuilder.add(shuffleNeededForNextStage)`?

I was thinking more about this. Is the end goal with the stage builder refactoring to allow each stage to do `queryDefBuilder.add(shuffleNeededForCurrentStage)` + `queryDefBuilder.add(work)`? (reversing the order, adding shuffling of current stage before work)

If my above understanding is correct, I feel this would be a bigger work item. Either way, I feel it's unrelated to the current refactoring PR, so this should be discussed separately in my opinion.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: I see we're using 2 fields from query context (in `WindowOperatorQueryFrameProcessor`):
1. `QueryKitUtils.CTX_TIME_COLUMN_NAME` (`__timeColumn`)
2. `DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY` (`sqlInsertSegmentGranularity`)

Shall I go ahead and raise a PR to pass `QueryContext` instead of the query itself?

I think it will be better to pass the whole query context instead of creating a configuration class with the above 2 fields, as it would allow flexibility in the future. Thoughts?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: With the latest PR changes where I added the `WindowStages` class, this is being done there now. Appreciate your inputs if this makes sense to you. Thanks!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: > if you need to create a new stage that have to be in the stages too

This is being done in `WindowStages` class now.

> you need to create a new stage when !currentStage.canAccept(of)

Have made the change.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: Can you elaborate on how it should be decided here?

Also, for this PR, I'm trying to keep this as a pure refactoring PR, so would like to avoid making any functional changes if that's okay.

**Review by kgyrtkirk — APPROVED** [review]

**Pull request merged**
